vm/devices/storage/disk_nvme/nvme_driver/src/driver.rs

```diff
@@ -680,10 +936,139 @@ impl<T: DeviceBacking> DriverWorkerTask<T> {
+    }
 }
 
 impl<T: DeviceBacking> InspectTask<WorkerState> for DriverWorkerTask<T> {
     fn inspect(&self, req: inspect::Request<'_>, state: Option<&WorkerState>) {
         req.respond().merge(self).merge(state);
     }
 }
+
+pub mod save_restore {
+    use super::*;
+
+    /// Save/restore state for NVMe driver.
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct NvmeDriverSavedState {
+        /// Admin queue state.
+        #[mesh(1)]
+        pub admin: Option<QueuePairSavedState>,
+        /// IO queue states.
+        #[mesh(2)]
+        pub io: Vec<QueuePairSavedState>,
+        /// Copy of the controller's IDENTIFY structure.
+        #[mesh(3)]
+        pub identify_ctrl: [u8; 4096],
+        /// Device ID string.
+        #[mesh(4)]
+        pub device_id: String,
+        /// Namespace data.
+        #[mesh(5)]
+        pub namespaces: Vec<crate::namespace::SavedNamespaceData>,
+        /// Queue size as determined by CAP.MQES.
+        #[mesh(6)]
+        pub qsize: u16,
+        /// Max number of IO queue pairs.
+        #[mesh(7)]
+        pub max_io_queues: u16,
+    }
+
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct QueuePairSavedState {
+        #[mesh(1)]
+        /// Which CPU handles requests.
+        pub cpu: u32,
+        #[mesh(2)]
+        /// Interrupt vector (MSI-X)
+        pub msix: u32,
+        #[mesh(3)]
+        pub sq_state: SubmissionQueueSavedState,
+        #[mesh(4)]
+        pub cq_state: CompletionQueueSavedState,
+        #[mesh(5)]
+        pub mem_len: usize,
+        #[mesh(6)]
+        pub pfns: Vec<u64>,
+        #[mesh(7)]
+        pub pending_cmds: PendingCommandsSavedState,
+    }
+
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct SubmissionQueueSavedState {
+        #[mesh(1)]
+        pub sqid: u16,
+        #[mesh(2)]
+        pub head: u32,
+        #[mesh(3)]
+        pub tail: u32,
+        #[mesh(4)]
+        pub committed_tail: u32,
+        #[mesh(5)]
+        pub len: u32,
+        #[mesh(6)]
+        pub pfns: Vec<u64>,
+    }
+
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct CompletionQueueSavedState {
+        #[mesh(1)]
+        pub cqid: u16,
+        #[mesh(2)]
+        pub head: u32,
+        #[mesh(3)]
+        pub committed_head: u32,
+        #[mesh(4)]
+        pub len: u32,
+        #[mesh(5)]
+        /// NVMe completion tag.
+        pub phase: bool,
+        #[mesh(6)]
+        pub pfns: Vec<u64>,
+    }
+
+    #[derive(Protobuf, Clone, Debug)]
+    #[mesh(package = "underhill")]
+    pub struct PendingCommandsSavedState {
+        #[mesh(1)]
+        pub commands: Vec<spec::Command>,
+        #[mesh(2)]
+        pub next_cid_high_bits: u16,
+    }
+
+    #[derive(Protobuf, Clone, Debug, FromBytes, FromZeroes)]
+    #[mesh(package = "underhill")]
+    pub struct PendingCommandSavedState {
```

Comment:
This is not used.

Reply:
Reverted to make it used again with a nice touch of ZeroCopyEncoding.